### PR TITLE
Reuse computed key memoiser

### PR DESCRIPTION
**Pull request by JLHwung**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | N.A.
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

In this PR we reuse the memoising expression of the computed class element key when transforming class fields. This optimizing trick can be applied to other plugins that requires memoising an expression, such as optional chaining and nullish coalescing as well.

I think eventually we should create a new helper for such purpose. For any given expression, an optimized memoising procedure can be
1) If the expression is a uid reference, treat it as a constant binding and thus it is ["static"](https://github.com/babel/babel/blob/380d186b77d32734c595652a40cbda5e4e109c5c/packages/babel-traverse/src/scope/index.ts#L557) within the scope
2) If the expression is an assignment `uid = expression`, treat `uid` as the memoiser id, and re-use the `uid = expression` assignment
3) Otherwise create a new `uid` for `expression` and construct the assignment `uid = expression`

Currently we only did the step 3 by `scope.maybeGenerateMemoised`, so Babel will create quite a few redundant memoisers if various transforms are applied to a given expression that needs memoising. For example, transforming a computed class accessor down to ES5 require 3 plugins: decorators, class properties and class transforms.

### CODE REVIEW

**Comment by JLHwung** [comment]
The challenge of porting these changes to `maybeGenerateMemoised` is that we have to change the return structure and the compatibility issue kicks in.

Currently `maybeGenerateMemoised` returns an Identifier when the expression is cached and null otherwise. In our use cases, we construct the assignment from the identifier and insert the assignment somewhere.

For point 1, `maybeGenerateMemoised` should return whether the expression is a UID, since a UID must have been assigned to some expression somewhere, we can treat the expression as static and use cloneNode elsewhere.

For point 2, the API should return whether the expression is of the format `uid = expression`. In this case, the API user should decide whether they want to move the whole assignment along with other created assignment to preserve the execution ordering, and insert the `uid` as references elsewhere.

I am wondering if we need a broader helper to cover these steps, e.g. `memoiseComputedKey` or `memoiseAccessorComputedKey`.

File: packages/babel-helper-create-class-features-plugin/src/misc.ts
Comment (liuxingbaoyu) [inline packages/babel-helper-create-class-features-plugin/src/misc.ts:151]: Sorry, I didn't understand why `hasUid` is used, can you explain it? Thanks!

Comment (JLHwung) [reply packages/babel-helper-create-class-features-plugin/src/misc.ts:151]: Good question. The `hasUid` checks whether the identifier is generated from `scope.generateUidIdentifier`, if it is, it is very likely generated by official Babel plugins, intended to be a memoiser. If not, it is from user code and we should handle it more conservatively.

**Review by liuxingbaoyu — COMMENTED** [review]
In the future maybe we can use `Set` or `WeakSet` to achieve this.

**Review by liuxingbaoyu — APPROVED** [review]

**Comment by nicolo-ribaudo** [comment]
I was going to suggest to re-use this new logic in other places, but looking out out tests I only found https://github.com/nicolo-ribaudo/babel/blob/3caeeb18833c4f04ccd3218af492e6fd3a072818/packages/babel-plugin-transform-class-properties/test/fixtures/nested-class/super-call-in-decorator/output.js where this would be useful (`_dec = _this = _super.call(this)`). However, it is rare enough that it's not worth blocking this PR on it (feel free to improve other similar cases, but also it's ok to decide that it's not necessary :) )

**Comment by JLHwung** [comment]
@nicolo-ribaudo Yes there are more optimizable paths. As you mentioned it helps the decorator memoization, we will see more such patterns once we start interleaving computed keys and decorator evaluations. For example, 

```js
let brandCheckC;

function captureAccess(fn) {
  brandCheckC = fn;
  return v => v
}
class C {
  [Symbol.iterator]() {}
  @captureClassBrandCheck(o => #p in o) #p
}
```

Currently Babel transforms it to
```js
var _computedKey, _dec, _init_p;
let brandCheckC;
function captureClassBrandCheck(fn) {
  brandCheckC = fn;
  return v => v;
}
_computedKey = _toPropertyKey(Symbol.iterator);
_dec = captureClassBrandCheck(o => #p in o);
class C {
  static {
    [_init_p] = _applyDecs(this, [[_dec, 0, "p", o => o.#p, (o, v) => o.#p = v]], [], 0, _ => #p in _).e;
  }
  [_computedKey]() {}
  #p = _init_p(this);
}
```

which is invalid because the private brand check `#p in o` is moved out of the class (See also part 5 of https://github.com/babel/babel/issues/16117). To address this we can mix computed keys assignment with decorators:

```js
var _computedKey, _dec, _init_p;
let brandCheckC;
function captureClassBrandCheck(fn) {
  brandCheckC = fn;
  return v => v;
}

class C {
  static {
    [_init_p] = _applyDecs(this, [[_dec, 0, "p", o => o.#p, (o, v) => o.#p = v]], [], 0, _ => #p in _).e;
  }
  [_computedKey = _toPropertyKey(Symbol.iterator), _dec = captureClassBrandCheck(o => #p in o), _computedKey]() {}
  #p = _init_p(this);
}
```

Now if we further transpile the result above to ES5, we should inform the class transform that `_computedKey` is already a "static" uid, so we don't have to memoize the key again.

Admittedly this case is not very practical, as users can just capture the brand check from the `access` property of the decorator context.

**Pull request merged**
